### PR TITLE
feat: installation lifecycle scripts

### PR DESCRIPTION
**Pull request by gemwuu**

### CODE REVIEW

File: test/postinstall.test.js [hunk 1]
```diff
@@ -35,28 +35,6 @@ describe('test/postinstall.test.js', () => {
       // prepare pass
       assert.equal(fs.readFileSync(path.join(root, 'node_modules', '.prepare.txt'), 'utf8'), 'success: prepare');
     });
-
-    it('should not run prepublish with production mode', async () => {
```
Comment (gemwuu) [inline test/postinstall.test.js:39]: [The npm@9 doesn't run prepublish only in production mode](https://github.com/cnpm/npm-cli/blob/latest/lib/commands/install.js#L153), even though [its document suggests this behavior](https://docs.npmjs.com/cli/v8/using-npm/scripts#prepare-and-prepublish), we follow the implementation.

Comment (fengmk2) [reply test/postinstall.test.js:39]: https://github.com/npm/npm/issues/10074#issuecomment-150068059 哈哈，当年我也+1 了

File: lib/lifecycle_scripts.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  'preinstall',
+  'install',
+  'postinstall',
+  'prepublish',
+  'preprepare',
+  'prepare',
+  'postprepare',
+];
+
+// scripts that shuold run in dependencies
```
Comment (fengmk2) [inline lib/lifecycle_scripts.js:21]: typo should

File: lib/lifecycle_scripts.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+    const log = util.format(
+      '> %s %s %s %s> %s',
+      displayName,
+      script,
+      root,
+      os.EOL,
+      cmd
+    );
+    console.info(log);
```
Comment (fengmk2) [inline lib/lifecycle_scripts.js:63]: console.log 支持 format 的，不需要单独使用 util.format

```js
console.info('> %s %s %s %s> %s', x, x, x)
```

**Comment by fengmk2** [comment]
@SocketSecurity ignore uglify-es@3.3.10
@SocketSecurity ignore uglify-js@3.17.4

**Review by fengmk2 — APPROVED** [review]
LGTM

**Pull request merged**